### DISCUSSION TITLE
Error handling in send_file

### DISCUSSION THREAD

**Discussion by kasium** [pallets/flask#5394] [Q&A] [answered]
Let's assume I use flask's send_file with an IO stream from S3. In theses streams it can happen that an error occurs.
Then I register an error handler for this error on the app, but it won't be called.

```py
from flask import Flask, send_file
from io import BytesIO

class S3IO(BytesIO):
    # just to showcase
    def read(self, __size: int | None = ...) -> str:
        raise RuntimeError

app = Flask(__name__)

@app.errorhandler(RuntimeError)
def handle_error():
    return "error", 400

@app.get("/")
def do():
    return send_file(S3IO(), "foo.txt")
```

Output
```
127.0.0.1 - - [26/Jan/2024 08:03:01] "GET / HTTP/1.1" 500 -
Error on request:
Traceback (most recent call last):
  File "venv/lib/python3.12/site-packages/werkzeug/serving.py", line 364, in run_wsgi
    execute(self.server.app)
  File "venv/lib/python3.12/site-packages/werkzeug/serving.py", line 327, in execute
    for data in application_iter:
  File "venv/lib/python3.12/site-packages/werkzeug/wsgi.py", line 365, in __next__
    data = self.file.read(self.buffer_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "foo.py", line 7, in read
    raise RuntimeError
RuntimeError
```

I'm not sure if it's intended that the error handler is not called. Should I instead overwrite the FileWrapper and handle exceptions there?

**Comment by davidism** [comment, marked as answer]
Streaming responses can't handle errors late. In WSGI (and I'm pretty sure ASGI too), once a response has started, the headers and status code have already been sent to the client, and the body has already been started. It's impossible to take that back and send something else instead.

**Reply by kasium** [reply]
Thanks for the hint
